src/layers/embeddings/qk_rotary_embeddings.rs

```diff
@@ -0,0 +1,270 @@
+    InvalidHeadWidth {
+        expected_head_width: usize,
+        head_width: usize,
+    },
+
+    #[snafu(display("Cannot calculate positions from cache"))]
+    PositionsFromCache { source: candle_core::Error },
+
+    #[snafu(display("Cannot apply rotary embeddings"))]
+    RotaryEmbeddings { source: RotaryEmbeddingsError },
+}
+
+#[derive(Debug)]
+pub struct QueryKeyRotaryEmbeddings {
+    rotary_width: usize,
+    rotary_embeds: RotaryEmbeddings,
+    pub head_width: usize,
+}
+
+impl QueryKeyRotaryEmbeddings {
+    /// Construct query-key rotary embeddings module.
+    ///
+    /// The rotary embedding will be precomputed for up to ``seq_len`` positions.
+    /// The embedding will be recomputed when a longer sequence is found in the
+    /// input.
+    ///
+    /// * `fraction` - Fraction of hidden width to apply rotary embeddings to.
+    ///   Must be in ``[0,1]``.
+    /// * `head_width` - Query and key attention head width.
+    /// * `seq_len` - Number of positions to initially precompute.
+    /// * `base` - The base used for theta_i (normally 10_000). Determines the cycle
+    ///    length of the embeddings.
+    /// * `device` - Device on which the module is to be initialized.
+    pub fn new(
+        fraction: f32,
+        head_width: usize,
+        seq_len: usize,
+        base: usize,
+        device: &Device,
+    ) -> Result<Self, QueryKeyRotaryEmbeddingsError> {
+        ensure!(
+            (0f32..=1f32).contains(&fraction),
+            InvalidRangeSnafu { fraction }
+        );
+
+        // Truncate to get the width in case it is fractional.
+        let rotary_width = (fraction * head_width as f32) as usize;
+        let rotary_embeds = RotaryEmbeddings::new(rotary_width, seq_len, base, device)
+            .context(RotaryEmbeddingsSnafu)?;
+
+        Ok(QueryKeyRotaryEmbeddings {
+            head_width,
+            rotary_embeds,
+            rotary_width,
+        })
+    }
+
+    /// Apply rotary embeddings to the query and key.
+    ///
+    /// Returns the input with the rotary embeddings applied. *Shape:*
+    /// `(batch_size, n_heads, seq_len, width_per_head)`
+    ///
+    /// * `query` - Query to apply the rotary embeddings to.
+    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
+    /// * `key` - Key to apply the rotary embeddings to.
+    ///   *Shape:* `(batch_size, n_heads, seq_len, width_per_head)`
+    /// * `positions` - Positions of the inputs. If no positions are
+    ///    provided, they are assumed to be `[0, seq_len)`.
+    ///    *Shape:* `(batch_size, seq_len)`
+    pub fn forward(
+        &self,
+        query: &Tensor,
+        key: &Tensor,
+        cache: Option<KeyValueCache>,
+        mut positions: Option<Tensor>,
+    ) -> Result<(Tensor, Tensor), QueryKeyRotaryEmbeddingsError> {
+        let (batch_size, n_heads, seq_len, head_width) =
+            query.shape().dims4().context(InvalidRankSnafu {
+                expected: 4usize,
```

Comment:
```suggestion
                expected: 4,
```
Can be inferred/omitted here, no?

Reply:
Oddly enough type inference fails in this case.